lib/src/Base/Func/EvaluationImplementation.cxx

```diff
@@ -162,22 +163,62 @@ Bool EvaluationImplementation::isActualImplementation() const
 
 /* Here is the interface that all derived class must implement */
 
+namespace {
+Sample evaluateSequential(const EvaluationImplementation & evaluation, const Sample & inSample)
+{
+  const UnsignedInteger size = inSample.getSize();
+  Sample outSample(size, evaluation.getOutputDimension());
+  // Simple loop over the evaluation operator based on point
+  // The calls number is updated by these calls
+  for (UnsignedInteger i = 0; i < size; ++i) outSample[i] = evaluation.operator()(inSample[i]);
+  return outSample;
+}
+
+/* Parallel evaluation */
+struct EvaluationPolicy
+{
+  const Sample & input_;
+  Sample & output_;
+  const EvaluationImplementation & evaluation_;
+
+  EvaluationPolicy( const Sample & input,
+                    Sample & output,
+                    const EvaluationImplementation & evaluation)
+    : input_(input)
+    , output_(output)
+    , evaluation_(evaluation)
+  {}
+
+  inline void operator()( const TBB::BlockedRange<UnsignedInteger> & r ) const
+  {
+    for (UnsignedInteger i = r.begin(); i != r.end(); ++i) output_[i] = evaluation_.operator()(input_[i]);
+  }
+
+}; /* end struct EvaluationPolicy */
+
+Sample evaluateParallel(const EvaluationImplementation & evaluation, const Sample & inSample)
+{
+  const UnsignedInteger size = inSample.getSize();
+  Sample result(size, evaluation.getOutputDimension());
+  const EvaluationPolicy policy( inSample, result, evaluation );
+  TBB::ParallelFor( 0, size, policy );
+  return result;
+}
+
+
+}
+
 /* Operator () */
 Sample EvaluationImplementation::operator() (const Sample & inSample) const
 {
   const UnsignedInteger inputDimension = getInputDimension();
   if (inSample.getDimension() != inputDimension) throw InvalidArgumentException(HERE) << "Error: the given sample has an invalid dimension. Expect a dimension " << inputDimension << ", got " << inSample.getDimension();
 
-  const UnsignedInteger size = inSample.getSize();
-  Sample outSample(size, getOutputDimension());
-  // Simple loop over the evaluation operator based on point
-  // The calls number is updated by these calls
-  for (UnsignedInteger i = 0; i < size; ++i) outSample[i] = operator()(inSample[i]);
+  Sample outSample((isParallel_ && !isCacheEnabled() && !isHistoryEnabled()) ? evaluateParallel(*this, inSample) : evaluateSequential(*this, inSample));
```

Comment:
@dbarbier Is there any reason why parallelism is activated only if cache and history are disabled? Also, what is the motivation of using free functions instead of methods for the parallel versions of the operator() method? Do we have to generalize this practice to other classes like Distribution?

Reply:
Most implementations of `operator()(Point&)` contains
```
    if (isHistoryEnabled_) {
      inputStrategy_.store(inP);
      outputStrategy_.store(outP);
    }
```
There is no guarantee that `inputStrategy_` and `outputStrategy_` remain consistent when there are multiple threads, one thread may add inP, then another thread adds inP and outP, then the first thread adds outP.

The only solution is to disable history, compute the whole Sample and fill in history if needed.  And it should be done for all classes implementing `operator()(const Sample&)`.

Cache seems to be unused, so I did not try hard to make it work, and decided to be conservative.

IMHO we should favor free function over method member when possible (ie. when do not need access to non-public members), yes.  For instance there is no need to recompile everything when you introduce free function, unlike a member.